WIP: the task system v0.1

kywch opened the pull request:
@jsuarez5341 Please focus on these 3 files on the task system
* `nmmo/task/task.py`
* `nmmo/tests/task/test_task_api_sampler.py`
* `nmmo/tests/task/test_base_predicates.py`

This PR got larger, partly because I also expanded the coverage of pylint and ended up fixing whole bunch of pylint fixes. Sorry for the confusion.

kywch committed:
implementing the task system

jsuarez5341 committed:
Update README.md

kywch committed:
set quantity=1 when creating non-stack items, ignoring provided quantity

kywch committed:
added tests for base predicates

kywch committed:
made pylint inspect all py files

kywch committed:
fixed pylint complaints

kywch committed:
listed all predicates to implement

jsuarez5341 committed:
Merge pull request #71 from CarperAI/v1.6

Merge latest Carper branch into a v1.7 release

jsuarez5341 committed:
Bump version to 1.7

jsuarez5341 committed:
Version bump and deps

jsuarez5341 committed:
Add pypi badge

kywch committed:
pulled change_spawn_pos, provide_item from the class

kywch committed:
WIP: adding base predicates, fine-tuning task API

kywch committed:
Merge branch 'NeuralMMO:v1.7' into event-log

kywch committed:
implmented all non event-based predicates

jsuarez5341 commented on nmmo/task/task.py:
Not always possible, but try to avoid <name>/<name>.py. as well as <name>.py:Name claesses. It creates either messy imports (i.e. from task.task import Task) or messy namespaces that over-import (i.e. in __init__ from task.task import Task or from task import task)

kywch replied:
I'll think over and discuss, but right now some things I can think of are ...

* for directory: curricula
* task.py: task_api.py
...

kywch replied:
Renamed task.py to task_api.py. Open to suggestions on the directory name.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+  cache_result: Dict # cache for result of team task evaluation
+
+  # - add extra info that is not in the datastore (e.g., spawn pos)
+  # - would IS_WITHIN, TICK, COUNT_DOWN be good here?
```
Sure, as convenience functions

kywch replied:
Yap, I will keep adding helper functions as needed.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+  def is_member(self, ent_id):
+    return ent_id in self.spawn_pos
+
+  def entity_or_none(self, ent_id):
```
Not a huge fan of _or_none naming, though I see why you've done it. @daveey, what are your thoughts on this?

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+    return result
+
+
+class GameStateGenerator:
```
Maybe GameStateView? Or GameStateExtractor?

kywch replied:
game_state.py --> move TeamGameState, GameStateGenerator into it

kywch replied:
renaming --- ent2team --> ent_to_team, something like this

kywch replied:
made game_state.py, moved TeamGameState and GameStateGenerator, and renaming all x2y ish to x_to_y

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+class GameStateGenerator:
+  def __init__(self, realm: Realm, config: Config):
+    self.config = deepcopy(config)
+    self.ent2team, self.team2ent = self._map_ent_team(realm)
```
Is this called multiple times or can it go in __init__?

kywch replied:
`_map_ent_team()` is used once, so it can go into `__init__()`

kywch replied:
Done.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+
+    tile_cols = TileState.State.attr_name_to_col
+
+    # CHECK ME: what gets return for eliminated teams? killed agents?
```
You shouldn't rely on obs from dead teams/agents, but you do get rewards/infos for that last step, and those usually have trajectory info

kywch replied:
Agreed. We should write tests for obs, rewards, dones, infos for the dead agents. 

I guess obs=None, rewards=-1, dones=1, infos=None. ... Am I right? Or, how should the required trajectory info look like?

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+  # in that case, we need a simple dataclass to pass remaining info
+
+
+class PredicateTask:
```
What about just Predicate?

kywch replied:
Will change the name

* `PredicateTask` -> `Predicate`
* `TaskForce` -> `Team` (though we need to scan it first)
* `Mission` -> `Task`

kywch replied:
Renamed to `Predicate`

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+  def __init__(self, *args):
+    self.name = self._task_name(args)
+
+  def _task_name(self, args):
```
Would __name__ make sense here, or do we want to keep this separate?

kywch replied:
My intention for separating it is to isolate the messy, tangential str conversion stuff from initing the class. We may even want to make it a util function later.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+    if self.name in team_gs.cache_result:
+      return team_gs.cache_result[self.name]
+
+  def _desc(self, class_type):
```
Should this be internal? (_)?

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+      "evaluate": self.__call__.__doc__
+    }
+
+  def description(self) -> Dict:
```
Ah I see ... hmm, is there a way to unify both?

kywch replied:
This was to deal with AND, OR, NOT, IMPLY, where I appended child node info, like:
    `desc.update({ 'desc_child': ["AND"] + [t.description() for t in self._tasks] })`

My idea of the description() -> Dict function is to retain all the task information, so that we can turn those into LLM prompts, for goal embedding, etc.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+  def description(self) -> Dict:
+    return self._desc("PredicateTask")
+
+  def __and__(self, other):
```
So this section is here to construct a static graph, as opposed to just calling the tasks and directly and/or/not'ing them. This may be the correct approach, but in that case, we should be explicit about why we need a static graph (why do we?)

kywch replied:
Hmm, I'm not sure I'm understanding your question. With the current implementation, one can use both ways
* `Success() &  ~ Failure() & Success()`
* `And(Success(), Not(Failure()), Success())`

While writing the tests, I'm pleasantly surprised that something like below works and now more like it, especially after making the Task class callable.
```
    task3 = Success() &  ~ Failure() & Success()
    self.assertTrue(task3(mock_gs, test_ent))
```

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+    return desc
+
+
+class TaskForce:
```
This is a Team. If we have other objects named Team elsewhere, we should figure that out

kywch replied:
Will change the name

kywch replied:
Renamed to `Team`. Don't see any problem so far.

kywch replied:
The name `Group` was suggested to replace `Team`. Also, it'd be great if `Group` should support any combination of different policies, which is not well supported by the nmmo env.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+
+
+# CHECK ME: this should produce the same map as the env realm spawn
+class TeamHelper:
```
Can we just merge this with Team?

kywch replied:
I agree with the direction. There are several places doing redundant team mapping, so merging team-ish things across TeamEnv and the task system would make things tidy.

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+    return TaskForce("All", list(self._ent2team.keys()))
+
+
+class Mission:
```
This is a Task, which makes more sense if we do PredicateTask -> Predicate

kywch replied:
Will change the name
* `PredicateTask` -> `Predicate`
* `TaskForce` -> `Team` (though we need to scan it first)
* `Mission` -> `Task`

kywch replied:
Renamed to `Task`

jsuarez5341 commented on nmmo/task/task.py:
```diff
@@ -0,0 +1,467 @@
+
+
+# pylint: disable=abstract-method
+class TaskWrapper(nmmo.Env):
```
Can this go in Task?

kywch replied:
I will think about this direction in the next iteration, perhaps also considering `TeamEnv`.

jsuarez5341 commented on tests/task/test_base_predicates.py:
test_predicates

kywch replied:
The number of tests (and thus codes) will be double, if we implement all the predicates. In that case we may want to break down this file.

kywch replied:
Renamed to `test_predicates.py`

jsuarez5341 commented on tests/task/test_base_predicates.py:
```diff
@@ -0,0 +1,314 @@
+    tick_success = 10
+    team_size_ge = 2
+    tick_timer = 9
+    death_note = [1, 2, 3]
```
lol

jsuarez5341 commented on tests/task/test_task_api_sampler.py:
Main thing needed here is some high-level tasks made up of real predicates, like in the code sample I shared with you. This would also be a good resource to iterate on as the end-user experience (including ELM)

kywch replied:
Yap, so far the tests are really testing the correctness of each predicate. We should provide more complex examples, perhaps using a task sampler.

We only have random task sampler now, but will make a heuristic task sampler after implementing all the predicates.

kywch committed:
renamed classes: Predicate, Team, Task

MarkHaoxiang committed:
testtask predicate close to entity

MarkHaoxiang committed:
Merge branch 'event-log' of https://github.com/CarperAI/nmmo-environment into event-log

MarkHaoxiang committed:
Completed (needs check) test_predicates.py

kywch committed:
renamed pylint.cfg so that pylint can run without --rcfile

kywch committed:
removed tests from path, fixed importing tests.testhelpers

kywch commented:
* Fell into a pylint rabbit hole. Renamed pylint.cfg to .pylintrc so that pylint can run without --rcfile flag.
* Modified scripts/pre-git-check.sh to explicitly run pylint on modified files only.
* Also removed tests from pytest python path, so that tests.testhelpers can be correctly imported.

kywch committed:
return a dummy lava tile when querying an out-of-index tile

kywch committed:
refactored task_api.py, moved classes into game_state.py predicate.py

kywch committed:
fixed some tests and added comments

kywch commented:
The changes pushed to here are:
* created nmmo/task/game_state.py and moved TeamGameState and GameStateGenerator into it
* created nmmo/task/predicate.py and moved Predicate, AND, OR, NOT, IMPLY into it
* modified the task modules to work with the new files
* fixed a few predicates based on the tests
* renamed vars/functions that had 2 in their name, replacing 2 with _to_
* added comments, edited a few predicate tests

So far all the non event-based predicates are done, and other predicates require implementing event-logging. Still, this would give an overview how the task system v0.1 would work. 

@jsuarez5341 @daveey @MarkHaoxiang please review, and let me know if it's good to merge into v1.6.

Remaining task system works include (1) event logging implementation, (2) 22 event-based predicates, (3) v0.2 discussion (e.g., streamlining team-related codes: TeamEnv, Team, TeamGameState, TeamHelper, ...). Please feel free to make suggestions, which we can incorporate in the next iterations.

Thank you!

kywch commented:
Sorry to mix pylint stuff in this PR ... it got really messy for review. 

Please focus on 
* `nmmo/task/task_api.py`
* `nmmo/task/game_state.py`
* `nmmo/task/predicate.py`
* `nmmo/tests/task/test_predicates.py`

Predicates are currently implemented in the below files, but we may want to put these in a different place.
* `nmmo/task/base_predicate.py`
* `nmmo/task/item_predicate.py`
* `nmmo/task/gold_predicate.py`
* `nmmo/task/event_predicate.py`

Another big naming issue is that is nmmo/`task` a good name, or which might be better. I am putting `curricula` out here.

daveey commented on .github/workflows/pylint-test.yml:
```diff
@@ -21,7 +21,7 @@ jobs:
       run: pytest
     - name: Analysing the code with pylint
-      run: pylint --rcfile=pylint.cfg --recursive=y nmmo tests
+      run: pylint --recursive=y nmmo tests
```
was this intentional?

kywch replied:
Yes, this was intentional.

I renamed pylint.cfg to .pylintrc, so that pylint can find it by default. This allows us to run the same pylint without the --rcfile flag. The `pre-git-check.sh` was also modified, and it works as expected.

daveey commented on nmmo/entity/player.py:
```diff
@@ -31,9 +31,9 @@ def __init__(self, realm, pos, agent, color, pop):
-    tasks = realm.config.TASKS
-    if tasks:
-      self.diary = Diary(self, tasks)
+    # tasks = realm.config.TASKS
```
should we just delete this? Or add a comment for why this is commented out

kywch replied:
I will also clean this part of the code in this iteration.

kywch replied:
I deleted this line of code but left the other diary related code.

There are a few places in the nmmo code that have `agent.diary`, `player.diary`, and `config.TASKS`, which was used with the original task system. In the next iteration, we may want to discuss how we can replace Diary/Achievement system with the new task system/event-log.

daveey replied:
Seems like we can just blow those away, they aren't being used for anything anymore

kywch replied:
I will clean these after making event-logging work.

daveey commented on nmmo/systems/item.py:
```diff
@@ -210,6 +210,9 @@ class Armor(Equipment, ABC):
   def __init__(self, realm, level, **kwargs):
     defense = realm.config.EQUIPMENT_ARMOR_BASE_DEFENSE + \
               level*realm.config.EQUIPMENT_ARMOR_LEVEL_DEFENSE
+    # ignore quantity for Armor, so that the default quantity of 1 is used
```
Why do we need to do this? When do you have quantity > 1?

kywch replied:
Other than stackable items, each item should occupy one inventory slot. But the current __init__ allows one to initiate an Armor with the quantity field of value 2 that take up one inventory space, for example. I came across this while making tests.

MarkHaoxiang replied:
@kywch There's a hard-to-reproduce bug where the assertion on line 594 of action.py fails. Can occasionally be found by test_monkey_action.py.

daveey commented on nmmo/systems/item.py:
```diff
@@ -371,6 +380,13 @@ def damage(self):
+
+  def __init__(self, realm, level, **kwargs):
+    # ignore quantity for Consumable, so that the default quantity of 1 is used
+    if 'quantity' in kwargs:
```
I'm concerned that we have to do this in a bunch of places, could you explain the problem? Maybe there's a cleaner solution?

kywch replied:
Thinking further, I think a cleaner way would be removing quantity arg and setting quantity=1 by default. There might be cases where we may want to init Ammo with quantity > 1, and it's only Ammo. So perhaps, I should add quantity only to Ammo and see if this could work.

daveey replied:
That sounds good, could you make that change? The way you have it is too hacky

kywch replied:
Will do after checking with Joseph on `droptable.py`

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,430 @@
+    super().__init__(num_tick)
+    self.num_tick = num_tick
+
+  def __call__(self, team_gs: TeamGameState, ent_id):
```
Can this take GameState rather than TeamGameState?

kywch replied:
As we discussed, we will go with a global GameState instead of the current TeamGameState.

kywch replied:
Done!

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,430 @@
+
+# CHECK ME: maybe this should be the default task?
+class LiveLong(Timer):
+  def __call__(self, team_gs, ent_id):
```
I would have expected ent_id to be passed in the constructor, but maybe I'm not understanding the use case

kywch replied:
I'm posting the code snippet that we discussed below.

kywch replied:
Predicates were revised to have the list of agents, the state of which are being evaluated.

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,430 @@
+    super().__call__(team_gs, ent_id)
+
+    # cache the result
+    team_gs.cache_result[self.name] = False
```
I don't understand how the caching works, it seems like you only use team_gs.cache_result if self.agents is empty? But what's the point of the cache?

kywch replied:
Cache will have to be implemented differently, based on the new design. I'll also do caching

kywch replied:
Caching inside the predicates are all removed and moved to the wrapper.

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,430 @@
+# TODO: the below line will be gone after implementation
```
can this go?

kywch replied:
I'll remove it.

kywch replied:
Removed!

daveey commented on nmmo/task/gold_predicate.py:
```diff
@@ -0,0 +1,88 @@
+    return False
+
+
+class TeamHoardGold(GoldPredicate):
```
Why can't these predicates take a TaskTarget and work on individuals or groups? Why do we want TeamHoardGold in addition to HoardGold?

kywch replied:
I'm posting the code snippet that we discussed below. We aim to remove Team predicates with this new design.

kywch replied:
All predicates are modified to work with a list of agents, so all the previous team predicates are removed.

daveey reviewed (changes requested):
Thanks for doing this, it's a huge change. I think I am still confused about TeamGameState and Predicates. Maybe we can do a call on Friday to discuss?

kywch commented:
Here is the new API design that we discussed today.

The goal here is to (1) clarify the relationships between predicates, tasks, reward, (2) use a global game state, (3) get rid of team-based predicates. 

I also want to mention that there are open issues, like more flexible "task assignement" and where/how to place the task system (core vs. wrapper), which we will continue to discuss.

```
class TeamEnv(nmmo.Env):
  # Predicate: states something about the subject
  #   CHECK ME: Team -> Group? or team is fine?   team_helper, TeamEnv, GroupEnv, ...
  
  class HoardGold(Predicate):
    def __init__(self, subject: Group, amount):
      pass

  t1 = HoardGold(Group([agent_0]), 5)
  t2 = HoardGold(group_helper.group(1), 10) # ~ team goal

  or one can define custom task functions: 
  def tf1(game_state) -> bool:
    pass
  
  def tf2(game_state, subject: Team) -> bool:
    pass

  t1, t2 should be callable.

  # task_assignment: which agent gets how much reward for doing what
  # this is initialized during reset, like env.reset(task_assignment)
  task_assignment = {
    agent_0: [(t1, 50), (t2, 100)],
    agent_1: [(tf1, -50)],
    agent_2: [(lambda gs: tf2(gs, team_helper.group(2)), 100)]
  }
  """
  task_assignment = {}

  def _compute_rewards(self, agents, dones):
    rewards = {}
    global_gs = {}

    for ent_id in agents:
      if ent_id in self.task_assignment:
        for task, at_stake in self.task_assignment.items():
          rewards[ent_id] += task(global_gs) * at_stake
```

kywch committed:
rewrote predicates to follow the new task api

kywch committed:
fixed pylint complaints

kywch commented:
Here is the summary of task-related changes.
* The task api and predicates were revised based on the code snippet we discussed
* The team-based game state was replaced with a global game state
* Each predicate has `subject`, the state of which are being evaluated to return true/false. The subject is completely different from task assignee, whom are rewarded when the task is evaluated as true.
* A small `TaskManager` class was added to help task assignment
* The codes that use population is all gone from the task. A new class `Group` allows any combination of agents. As a result, `TeamHelper` is not being used at all. We may want to clean up team/population in the next iteration.
* Team-based predicates are all gone. The same predicates are used for both a single agent (treated as a group of one) and any list of agents. Thus, the number of planned predicates is now 28, and currently 14 non-event-based predicates have been implemented.
* Result caching is done in the wrapper, and the previous caching codes in each predicates are all removed, making the predicate code simpler.
* More test cases were added.

@daveey @jsuarez5341 please review and feel free to make suggestions. Meanwhile, I'll work on David's other comments (e.g., item quantity, achievement). Thanks!

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,166 @@
+    return len(alive_subject) == len(self.subject)
+
+
+class RestInPeace(Predicate):
```
Cute, but let's call it AllDead

kywch replied:
renamed.

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,166 @@
+    return len(alive_subject) == 0
+
+
+class SearchTile(Predicate):
```
CanSeeTile

kywch replied:
renamed

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,166 @@
+    return result
+
+
+class SearchAgent(Predicate):
```
CanSeeAgent

kywch replied:
renamed.

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,166 @@
+    return np.any(flt_row & flt_col)
+
+
+class GoDistance(Predicate):
```
DistanceTraveled

kywch replied:
renamed.

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,166 @@
+    return sum_dist >= self._dist
+
+
+class StayClose(Predicate):
```
AllMembersWithinRange

kywch replied:
renamed.

daveey commented on nmmo/task/base_predicate.py:
```diff
@@ -0,0 +1,166 @@
+    return max(max(rows)-min(rows), max(cols)-min(cols)) <= self._dist
+
+
+def skill_to_str(skill: Skill.Skill):
```
Can we define a skill.Melee.name() and use that? Doing string manipulation on class name is super hacky

kywch replied:
haha, there was an easier way without the hacky function -- `skill.__name__.lower()`

so just removed that function and used `skill.__name__.lower()`

daveey commented on nmmo/task/game_state.py:
```diff
@@ -0,0 +1,94 @@
+
+@dataclass
+class GameState:
+  tick: int
```
current_tick

daveey commented on nmmo/task/game_state.py:
```diff
@@ -0,0 +1,94 @@
+  alive_agents: List # of alive agents' ent_id
+  env_obs: Dict[int, Observation] # this should include only alive agents
+
+  entity_cols: Dict # attr_to_col
```
Can we just use Entity.State.attr_to_col_name? No point in copying that into every GameState when it can just be imported

daveey replied:
You can just add 

EntityAttr = EntityState.State.attr_name_to_col

To the top of the file

kywch replied:
defined EntityAttr, ItemAttr, TileAttr at the top of the file and using these, instead of passing these via game state.

daveey commented on nmmo/task/game_state.py:
```diff
@@ -0,0 +1,94 @@
+
+    return GameState(
+      tick = deepcopy(realm.tick),
+      config = deepcopy(self.config),
```
deep copy seems expensive. why can't we just provide a reference to the config?

kywch replied:
Removed deepcopy for the config, since it was deep-copied when the game state generator was initiated. Messing with it won't affect the game.

The other things that are deep-copied are: env_obs, spawn_pos, entity_data, and item_data (these numpy tables are copied via astype(np.int16) function) 
* Messing with env_obs can affect the next step, so it'd be good to protect it.
* Messing with spawn_pos can affect the `DistanceTraveled` predicate. I don't think it's high risk, but since it's small, so why not?

kywch replied:
Chatgpt taught me `@dataclass(frozen=True)`, which can keep everything read only without deep copy.

gs.cache_result is still available to write via `MutableMapping`. Cool.

daveey commented on nmmo/task/predicate.py:
```diff
@@ -0,0 +1,256 @@
+class Group:
+  def __init__(self, agents:List[int], name:str=None):
+    assert len(agents) > 0, "Team must have at least one agent"
+    self.name = name if name else f"Agent{str(agents).replace(' ', '')}"
```
How about f"Agents("{",".join(agents)}")"

kywch replied:
Changed to `f"Agent({','.join([str(e) for e in agents])})"`, which produces something like: 'Agent(4,5,6,7)'

It seems that we also would a group of one agent a lot, so I'm leaning toward Agent (vs. Agents)

daveey commented on nmmo/task/predicate.py:
```diff
@@ -0,0 +1,256 @@
+  def subject(self) -> List[int]:
+    return self._subject
+
+  def _task_name(self, subject:Group, args):
```
it's no-longer a task

kywch replied:
Renamed to `_make_name()`. 

Also replaced hacky string manipulations with `arg.__name__` and `arg.__class__.__name__`

daveey commented on nmmo/task/predicate.py:
```diff
@@ -0,0 +1,256 @@
+  def subject(self) -> List[int]:
+    return self._subject
+
+  def _task_name(self, subject:Group, args):
```
minor formatting, but you need a space after ": " here and in a bunch of places. If lint doesn't catch it, could you add a rule?

kywch replied:
Oh I didn't know that. I'm fixing it as I see it. 

I looked into pylint, and unfortunately, pylint removed `bad-whitespace` check from 2.6.0. 
https://pylint.readthedocs.io/en/latest/whatsnew/2/2.6/full.html#what-s-new-in-pylint-2-6-0

`bad-continuation and bad-whitespace have been removed, black or another formatter can help you with this better than Pylint`

daveey commented on nmmo/task/predicate.py:
```diff
@@ -0,0 +1,256 @@
+    desc.update({ 'desc_child': ["NOT", self._task.description] })
+    return desc
+
+class IMPLY(Predicate):
```
When do we think we'd use this? Since it can be done with NOT and OR, it seems unnessary

kywch replied:
looked fancy, but removing it.

kywch replied:
hmm, I wasn't sure too but looked fancy so I kept it... Why don't we try with ELM and see what it can make out of IMPLY?

MarkHaoxiang committed:
Removed traces of diary and old task system.

MarkHaoxiang committed:
Moved skill str definition

MarkHaoxiang committed:
GroupView - a easier to use view of GameState for a group with computational optimization.

MarkHaoxiang committed:
wip: edit predicates to use new helper class

kywch committed:
renamed predicates, removed deepcopy from gs, etc

kywch commented:
@daveey, I incorporated your feedback, except the item/quantity, which I'm working on now.

Please review and let me know if you have any suggestions. Thanks!

kywch commented:
Item quantity-related commit is in the separate branch (`item-quantity`): https://github.com/CarperAI/nmmo-environment/commit/025c2201a913e05bf90a342778be6c3a4a52d8dd

Removed quantity from Item.__init__() and `droptable.py`, and it doesn't seem to interfere with the current unit tests. 

Ok, so the `item-quantity` branch was merged to here.

MarkHaoxiang committed:
Base predicate modification

MarkHaoxiang committed:
Pylinted. Finished modification of item predicates.

kywch committed:
removed quantity from Item.__init__() (#34)

MarkHaoxiang committed:
Merge branch 'mark-task' into event-log

MarkHaoxiang committed:
Fixed errors from merge.

kywch commented:
Tooling related changes were pushed separately, and the task system v0.1 is obsolete.

Closing this PR without merge.
